openrouteservice/src/test/java/org/heigit/ors/matrix/core/CoreMatrixTest.java

```diff
@@ -0,0 +1,1167 @@
+
+        CoreMatrixAlgorithm algorithm = new CoreMatrixAlgorithm();
+        EdgeFilterSequence edgeFilterSequence = new EdgeFilterSequence();
+
+        CoreTestEdgeFilter restrictedEdges = new CoreTestEdgeFilter();
+        restrictedEdges.add(9);
+        edgeFilterSequence.add(restrictedEdges);
+        CHGraph g = contractGraph(graphHopperStorage, edgeFilterSequence);
+
+        MatrixLocations sources = new MatrixLocations(1);
+        sources.setData(0, 0, null);
+        MatrixLocations destinations = new MatrixLocations(1);
+        destinations.setData(0, 7, null);
+
+        MatrixRequest matrixRequest = new MatrixRequest();
+        matrixRequest.setMetrics(MatrixMetricsType.DISTANCE);
+
+        Weighting turnWeighting = new TurnWeighting(weighting, HelperORS.getTurnCostExtensions(graphHopperStorage.getExtension()), 0);
+        algorithm.init(matrixRequest, g, carEncoder, weighting, new CoreTestEdgeFilter());
+        MatrixResult result = null;
+        try{
+            result = algorithm.compute(sources, destinations, MatrixMetricsType.DISTANCE);
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+        assertEquals(6.0, result.getTable(MatrixMetricsType.DISTANCE)[0], 0);
+
+    }
+
+    @Test
+    public void testDownwardPassFasterUTurn() {
+        GraphHopperStorage graphHopperStorage = ToyGraphCreationUtil.createTwoWayGraph(createGHStorage());
+        addRestrictedTurn(graphHopperStorage, 2, 3, 3);
+
+        CoreMatrixAlgorithm algorithm = new CoreMatrixAlgorithm();
+        EdgeFilterSequence edgeFilterSequence = new EdgeFilterSequence();
+        edgeFilterSequence.add(new TurnRestrictionsCoreEdgeFilter(carEncoder, graphHopperStorage));
+        CoreTestEdgeFilter restrictedEdges = new CoreTestEdgeFilter();
+        restrictedEdges.add(10);
+        restrictedEdges.add(11);
+        edgeFilterSequence.add(restrictedEdges);
+        CHGraph g = contractGraph(graphHopperStorage, edgeFilterSequence);
+
+        MatrixLocations sources = new MatrixLocations(1);
+        sources.setData(0, 7, null);
+        MatrixLocations destinations = new MatrixLocations(2);
+        destinations.setData(0, 5, null);
+        destinations.setData(1, 7, null);
+
+        MatrixRequest matrixRequest = new MatrixRequest();
+        matrixRequest.setMetrics(MatrixMetricsType.DISTANCE);
+
+        Weighting turnWeighting = new TurnWeighting(weighting, HelperORS.getTurnCostExtensions(graphHopperStorage.getExtension()), 0);
+        algorithm.init(matrixRequest, g, carEncoder, turnWeighting, new CoreTestEdgeFilter());
+        MatrixResult result = null;
+        try{
+            result = algorithm.compute(sources, destinations, MatrixMetricsType.DISTANCE);
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+        }
+        //TODO this would be the correct result, but the algorithm cannot provide this currently.
```

Comment:
It's an interesting point. With infinite u turn costs, this result is wrong. But it cannot be adressed currently by the algorithm. It is also an extremely rare case (I only found one in all my tests). 
But it is reproducable. See [Update of turn restricted node in downward pass](https://confluence.gistools.geog.uni-heidelberg.de/pages/viewpage.action?spaceKey=GC&title=Persisting+problems) What do you think?

Reply:
> This can probably only be resolved by using edge based calculations outside of the core. 

I don't seem to observe this issue with CALT routing which doesn't use edge-based calculations outside of the core. Aren't u-turns anyway blocked outside of the core?

I'm not sure whether this would help. CALT does not use

Reply:
That is because Core ALT is bidirectional, there is never a downwards pass. Core Matrix is one directional, it goes to the core and then inside and down (at the same time). The downwards pass overwrites a correct value with a better, but incorrect one. You can check the graph that is created in the test, it is written exactly for this problem

Reply:
Its not actually about u turns, just turn restriction in general ,sorry

Reply:
I somehow cannot get my head around how the weight of a turn-restricted node which is a core node can be altered by the downward pass which runs outside (?) of the core. It sounds counterintuitive to me.

Reply:
I guess we have to live with this for now. 🤷

Reply:
```
        // 0<----------<-1
        // |             |
        // 2             |
        // | R           |
        // 3---4---5     |
        // |             |
        // 6-----7-------8
        // |
        // 9
        g.edge(0, 2, 1, false); //0
        g.edge(1, 0, 1, false); //1
        g.edge(2, 3, 1, false); //2
        g.edge(3, 4, 1, false); //3
        g.edge(4, 5, 1, true); //4
        g.edge(3, 6, 1, true); //5
        g.edge(7, 8, 1, true); //6
        g.edge(6, 9, 1, true); //7
        g.edge(6, 7, 10, true); //8
        g.edge(8, 1, 1, true); //9
        g.edge(8, 1, 1, true); //10 Just to put 8, 1 and 0 in core
        g.edge(1, 0, 1, false); //11  Just to put 8, 1 and 0 in core
```
We go from 7 to 5. The way from 2 via 3 to 4 is restricted. But still, I can reach node 3 via the way `7-6-3`, which is longer than the right side loop via 8. But then I cannot continue because of the restriction. Instead, the best path is found by going down from the core from node 3 to node 6. This will update the weight of node 6 to be smaller (before it was 10, now it is 6), and update node 6 outside the core. This is correct behavior. 
But now, the path `7-6-3` is not weight 11 anymore, but instead has an updated entry in it, which makes it look like it is only weight 7.
Basically two paths, `7-6-3-4-5` and `7-8-1-0-2-3-6` get "melted" into one, because the entry at node 6 is node-based and can only store one value, the best one for node 6. Makes sense?
That is why it can only be adressed with edge based outside core. We need to store two values for node 6 instead of one.